core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -188,6 +191,17 @@ public ZenDiscovery(Settings settings, ThreadPool threadPool, TransportService t
             DISCOVERY_REJOIN_ACTION_NAME, RejoinClusterRequest::new, ThreadPool.Names.SAME, new RejoinClusterRequestHandler());
     }
 
+    // protected to allow overriding in tests
+    protected ZenPing newZenPing(Settings settings, ThreadPool threadPool, TransportService transportService,
+                                 UnicastHostsProvider hostsProvider) {
+        return new UnicastZenPing(settings, threadPool, transportService, hostsProvider);
+    }
+
+    // TODO: this shoudln't be public, clean up the zen ping interface somehow to better allow clearing for testing?
```

Comment:
well, we can call it `ZenTestDiscovery` :) - my point is more that then we can move the  `getZenPing` to the test only variant and not have it public on `ZenDiscovery`. There were also times where we considered mocking elect master service where this test-only discovery would have been useful.